admin/command_runner.go

```diff
@@ -0,0 +1,306 @@
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	r.logger.Info().Msg("admin server starting up")
+
+	listener, err := net.Listen("unix", r.grpcAddress)
+	if err != nil {
+		return fmt.Errorf("failed to listen on admin server address: %w", err)
+	}
+
+	grpcServer := grpc.NewServer()
+	pb.RegisterAdminServer(grpcServer, NewAdminServer(r.commandQ))
+
+	r.workersStarted.Add(1)
+	r.workersFinished.Add(1)
+	go func() {
+		defer r.workersFinished.Done()
+		r.workersStarted.Done()
+
+		if err := grpcServer.Serve(listener); err != nil {
+			r.logger.Err(err).Msg("gRPC server encountered fatal error")
+			r.errors <- err
+		}
+	}()
+
+	// Register gRPC server endpoint
+	mux := runtime.NewServeMux()
+	opts := []grpc.DialOption{grpc.WithInsecure()}
+
+	err = pb.RegisterAdminHandlerFromEndpoint(ctx, mux, "unix:///"+r.grpcAddress, opts)
+	if err != nil {
+		return fmt.Errorf("failed to register http handlers for admin service: %w", err)
+	}
+
+	httpServer := &http.Server{
+		Addr:      r.httpAddress,
+		Handler:   mux,
+		TLSConfig: r.tlsConfig,
+	}
+
+	r.workersStarted.Add(1)
+	r.workersFinished.Add(1)
+	go func() {
+		defer r.workersFinished.Done()
+		r.workersStarted.Done()
+
+		// Start HTTP server (and proxy calls to gRPC server endpoint)
+		var err error
+		if r.tlsConfig == nil {
+			err = httpServer.ListenAndServe()
+		} else {
+			err = httpServer.ListenAndServeTLS("", "")
+		}
+
+		if err != nil && !errors.Is(err, http.ErrServerClosed) {
+			r.logger.Err(err).Msg("HTTP server encountered error")
+			r.errors <- err
+		}
+	}()
+
+	r.workersStarted.Add(1)
+	r.workersFinished.Add(1)
+	go func() {
+		defer r.workersFinished.Done()
+		r.workersStarted.Done()
+
+		<-ctx.Done()
+		r.logger.Info().Msg("admin server shutting down")
+
+		grpcServer.Stop()
+		close(r.commandQ)
+
+		if httpServer != nil {
+			shutdownCtx, shutdownCancel := context.WithTimeout(context.Background(), CommandRunnerShutdownTimeout)
+			defer shutdownCancel()
+
+			if err := httpServer.Shutdown(shutdownCtx); err != nil {
+				r.logger.Err(err).Msg("failed to shutdown http server")
+				r.errors <- err
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (r *CommandRunner) processLoop(ctx context.Context) {
+	defer func() {
+		// cleanup uncompleted requests from the command queue
+		for command := range r.commandQ {
+			close(command.responseChan)
+		}
+
+		r.workersFinished.Done()
+	}()
+
+	r.workersStarted.Done()
+
+	for {
+		select {
+		case command := <-r.commandQ:
+			r.logger.Info().Str("command", command.command).Msg("received new command")
+
+			var err error
+
+			if validator := r.getValidator(command.command); validator != nil {
+				if validationErr := validator(command.data); validationErr != nil {
+					err = status.Error(codes.InvalidArgument, validationErr.Error())
+					goto sendResponse
+				}
+			}
+
+			if handler := r.getHandler(command.command); handler != nil {
+				// TODO: we can probably merge the command context with the worker context
```

Comment:
That's an excellent idea, which is probably worth tracking in an issue, and has links with #1275 (or at least the context-derived variant thereof I suggest in #1308 ).

Reply:
https://github.com/onflow/flow-go/issues/1314